ingester/user_state.go

```diff
@@ -183,17 +188,54 @@ func (u *userState) unlockedGet(metric model.Metric, maxSeries int) (model.Finge
 	// all proceed to add a new series. This is likely not worth addressing,
 	// as this should happen rarely (all samples from one push are added
 	// serially), and the overshoot in allowed series would be minimal.
-	if u.fpToSeries.length() >= maxSeries {
+	if u.fpToSeries.length() >= cfg.MaxSeriesPerUser {
 		u.fpLocker.Unlock(fp)
 		return fp, nil, util.ErrUserSeriesLimitExceeded
 	}
 
+	metricName, err := util.ExtractMetricNameFromMetric(metric)
+	if err != nil {
+		u.fpLocker.Unlock(fp)
+		return fp, nil, err
+	}
+
+	// The same race can happen here as with series-per-user above.
+	if u.numSeriesInMetric(metricName) >= cfg.MaxSeriesPerMetric {
```

Comment:
you could roll `numSeriesInMetric` and `incNumSeriesInMetric` into one call (`canAddSeries`) under one lock, and both simplify the code and remove the race..